File: app/src/main/java/com/silentgames/silent_planet/model/entities/EntityType.java
package com.silentgames.silent_planet.model.entities;

import com.silentgames.silent_planet.model.entities.ground.PlayersOnCell;
import com.silentgames.silent_planet.model.entities.space.SpaceShip;

/**
 * Created by gidroshvandel on 09.07.16.
 */

public class EntityType extends EntityTypeEx {

    private PlayersOnCell playersOnCell;

    private SpaceShip spaceShip;


    public EntityType(PlayersOnCell player) {
        this.playersOnCell = player;
        setAll(player.getPlayerList().get(0));
    }

    public EntityType(SpaceShip spaceShip) {
        this.spaceShip = spaceShip;
        setAll(spaceShip);
    }

    public SpaceShip getSpaceShip() {
        return spaceShip;
    }

    public void setSpaceShip(SpaceShip spaceShip) {
        this.spaceShip = spaceShip;
    }

    public PlayersOnCell getPlayersOnCell() {
        return playersOnCell;
    }

    public void setPlayersOnCell(PlayersOnCell playerList) {
        this.playersOnCell = playerList;
    }
}

File: app/src/main/java/com/silentgames/silent_planet/model/entities/EntityType.kt
package com.silentgames.silent_planet.model.entities

import android.content.Context
import com.silentgames.silent_planet.model.effects.Effect
import com.silentgames.silent_planet.model.fractions.Fractions

/**
 * Created by gidroshvandel on 09.07.16.
 */

abstract class EntityType(
        val context: Context,
        override var fraction: Fractions,
        override var crystals: Int = 0,
        override var isDead: Boolean = false,
        val effects: MutableList<Effect> = mutableListOf()
) : EntityTypeProperties {

    override val id: String get() = name + "_" + description + "_" + fraction.fractionsType.name

    fun addEffect(effect: Effect) {
        effects.add(effect)
    }

    fun removeEffect(effect: Effect) {
        effects.remove(effect)
    }

    fun addCrystals(crystals: Int) {
        this.crystals = this.crystals + crystals
    }

    fun removeCrystals(crystals: Int) {
        val result = this.crystals - crystals
        this.crystals = if (result >= 0) result else 0
    }

    inline fun <reified T : Effect> hasEffect() = getEffect<T>() != null

    inline fun <reified T : Effect> getEffect() = effects.filterIsInstance<T>().firstOrNull()

    override fun equals(other: Any?): Boolean {
        return if (other is EntityType) {
            other.id == id
        } else {
            super.equals(other)
        }
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + description.hashCode()
        result = 31 * result + fraction.fractionsType.hashCode()
        return result
    }
}
